### PR TITLE
[gitlab] Artifact test suite logs on failure.

### PR DESCRIPTION
**Pull request by SkySkimmer**
An sample result can be found at https://gitlab.com/SkySkimmer/coq/builds/17510262/artifacts/download (with an artificial failure in output/Arguments).

### CODE REVIEW

**Comment by ejgallego** [comment]
That's something I've been wanting for ages, however it would be great if we could:
- only save the logs of failed tests
- make them be displayed in the log, so this also works for Travis.

**Comment by maximedenes** [comment]
Oh, so gitlab has support for artifacts. That would be extremely useful to generate our packages. Travis doesn't have it...

**Comment by Zimmi48** [comment]
@SkySkimmer is giving us each day more reasons for moving to GitLab.com

**Comment by ejgallego** [comment]
¿? https://docs.travis-ci.com/user/uploading-artifacts/

**Comment by maximedenes** [comment]
Yes, since Travis doesn't support artifacts, they list external services to which they can interface. In fact, I was planning to use bintray. But this is still way less convenient than an integrated support for artifacts. In fact, it is the only thing I miss from our former Jenkins setup.

**Comment by SkySkimmer** [comment]
Gitlab is supposed to have some jenkins integration but I don't know anything about jenkins so I don't know how far it goes.
Also I don't know how well OSX/windows testing works with gitlab.

**Comment by ejgallego** [comment]
Travis does support artifacts, what it does not provide is hosting for them.

**Comment by maximedenes** [comment]
> Travis does support artifacts, what it does not provide is hosting for them.

Which is what I call not supporting them. Otherwise no need for support, I can run an `scp` command myself.

**Comment by Zimmi48** [comment]
@SkySkimmer GitHub is also supposed to have support for Jenkins, but that never worked.

**Comment by ejgallego** [comment]
> Otherwise no need for support, I can run an scp command myself.

Incorrect on two accounts (where would the key come from, and how would you manage the logic used by the `artifact` keyword).

**Comment by maximedenes** [comment]
> where would the key come from

Hidden variables, as documented by Travis. In fact, how do you think authentication to S3 or other external services work?

> how would you manage the logic used by the artifact configuration option

we don't need any of that AFAICT

**Comment by SkySkimmer** [comment]
The major difference is that gitlab artifacts work in forks and pull requests (except they're called merge requests) but travis secret variables (and travis artifact uploading) are intra repository only.

**Comment by SkySkimmer** [comment]
@ejgallego 
>That's something I've been wanting for ages, however it would be great if we could:
>-    only save the logs of failed tests
>-   make them be displayed in the log, so this also works for Travis.

Like this?

**Comment by SkySkimmer** [comment]
Something weird is going on with travis, only the appveyor automatic test is being run for this PR.

**Comment by ejgallego** [comment]
Looks good thanks.

**Comment by SkySkimmer** [comment]
Right I didn't test the case where no test fails.

**Comment by SkySkimmer** [comment]
It should work now.
Example with artificial failure: https://travis-ci.org/SkySkimmer/coq/jobs/237932526

**Comment by maximedenes** [comment]
Thanks, this is very useful!

File: Makefile.build [hunk 1]
```diff
@@ -454,7 +454,6 @@ check: validate test-suite
 test-suite: world $(ALLSTDLIB).v
 	$(MAKE) $(MAKE_TSOPTS) clean
 	$(MAKE) $(MAKE_TSOPTS) all
-	$(MAKE) $(MAKE_TSOPTS) report
```
Comment (maximedenes) [inline Makefile.build]: Why was this target removed from here?

Comment (SkySkimmer) [inline Makefile.build]: See 38d3aa9: test suite make all includes make report so it was printing the report twice.
For instance if we pick some random old build https://travis-ci.org/coq/coq/jobs/216784644:
>make[3]: Entering directory `/home/travis/build/coq/coq/test-suite`
BUILDING SUMMARY FILE
NO FAILURES
make[3]: Leaving directory `/home/travis/build/coq/coq/test-suite`
make[2]: Leaving directory `/home/travis/build/coq/coq/test-suite`
make -C test-suite -s VERBOSE= report
make[2]: Entering directory `/home/travis/build/coq/coq/test-suite`
BUILDING SUMMARY FILE
NO FAILURES
make[2]: Leaving directory `/home/travis/build/coq/coq/test-suite`

Comment (maximedenes) [inline Makefile.build]: Ok, thanks for the explanation!

**Pull request merged**